services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,425 @@
+
+    private TenantInformationService tenantInformationService = new NoopTenantInformationService();
+
+    private DownstreamSenderFactory downstreamSenderFactory;
+
+    private Future<HonoConnection> connectionAttempt;
+
+    private Vertx vertx;
+
+    private AutoProvisionerConfigProperties config;
+
+    @Override
+    public final Future<Void> start() {
+        // Required since every endpoint may call the start() method of its referenced service leading to multiple calls
+        // to start(). This results in an exception in the factory's connect() method.
+        synchronized (this) {
+            if (connectionAttempt == null) {
+                connectionAttempt = downstreamSenderFactory.connect().map(connection -> {
+                    LOG.info("connected to AMQP network");
+                    if (vertx == null && connection != null) {
+                        vertx = connection.getVertx();
+                    }
+                    return connection;
+                }).recover(t -> {
+                    LOG.warn("failed to connect to AMQP network", t);
+                    return Future.failedFuture(t);
+                });
+            }
+        }
+
+        return connectionAttempt.mapEmpty();
+    }
+
+    @Override
+    public final Future<Void> stop() {
+        final Promise<Void> result = Promise.promise();
+        downstreamSenderFactory.disconnect(result);
+        return result.future();
+    }
+
+    /**
+     * Sets the vert.x instance.
+     *
+     * @param vertx The vert.x instance.
+     * @throws NullPointerException if vert.x is {@code null}.
+     */
+    @Autowired
+    public final void setVertx(final Vertx vertx) {
+        this.vertx = Objects.requireNonNull(vertx);
+    }
+
+    /**
+     * Sets the factory to use for creating a client for the AMQP Messaging Network.
+     *
+     * @param factory The factory.
+     * @throws NullPointerException if the factory is {@code null}.
+     */
+    @Autowired
+    public final void setDownstreamSenderFactory(final DownstreamSenderFactory factory) {
+        this.downstreamSenderFactory = Objects.requireNonNull(factory);
+    }
+
+    /**
+     * Sets the {@link DeviceManagementService} to use.
+     *
+     * @param deviceManagementService The service to set.
+     *
+     * @throws NullPointerException if the service is {@code null}.
+     */
+    @Autowired
+    public void setDeviceManagementService(final DeviceManagementService deviceManagementService) {
+        this.deviceManagementService = Objects.requireNonNull(deviceManagementService);
+    }
+
+    /**
+     * Sets the {@link DeviceRegistrationInformationService} to use.
+     *
+     * @param deviceRegistrationInformationService The service to set.
+     *
+     * @throws NullPointerException if the service is {@code null}.
+     */
+    @Autowired
+    public void setDeviceRegistrationInformationService(final DeviceRegistrationInformationService deviceRegistrationInformationService) {
```

Comment:
if we already have a `DeviceManagementService` then why do we also need a `DeviceRegistrationInformationService`? Couldn't we simply use `DeviceManagementService.readDevice()`?

Reply:
Hmm. Good question. When refactoring I noticed that the implementation of the _getRegistrationInformation_ method is slightly different from _readDevice_.

I think (at least for MongoDb) it can be replaced but I didn't want to add this to the scope of this PR.

Reply:
In particular, `DeviceManagementService.readDevice()` returns a `Device` object instance instead of just Json which should actually make implementation of the *sendDelayedAutoProvisioningNotificationIfNeeded* method easier as there is no need to parse Json ...

So, unless there is a more specific reason than
> the implementation of the getRegistrationInformation method is slightly different from readDevice

I think we should remove this field ...

Reply:
Done.